install-docker.sh

```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+set +eux
+
+# https://docs.docker.com/engine/install/ubuntu/
+# Add Docker's official GPG key:
+sudo apt-get update
+sudo apt-get install -y ca-certificates curl
+sudo install -m 0755 -d /etc/apt/keyrings
+sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+sudo chmod a+r /etc/apt/keyrings/docker.asc
+
+# Add the repository to Apt sources:
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+sudo apt-get update
+
+# Install docker
+sudo apt-get -y install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
+
+# Install NVidia docker engine runtime
+# https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html
+curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
+  && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
+    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
+    sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
+sudo apt-get update
+sudo apt-get install -y nvidia-container-toolkit
+
+# Rootless mode
+sudo apt-get install -y dbus-user-session
+sudo systemctl stop docker docker.socket
+sudo systemctl disable --now docker.service docker.socket
+
+sudo dockerd-rootless-setuptool.sh install
+dockerd-rootless-setuptool.sh install
```

Comment:
In the other PR, you explained that we use rootless mode because we're using dind, but this is not very clear to me (or maybe I misunderstood something): This script installs docker in the host machine, so we can simply use root-full mode, right?

Reply:
And also one thing not very clear is that how many dockerd instance will be created? We're using only one dockerd provided by host, then mount it into container right?

Reply:
No we use rootless docker on the VM. No need to have root docker service.

Reply:
If we're only using the host's dockerd and mount the docker.sock inside the container, then it is Docker outside of Docker (DooD) and not DinD.

Sorry if my question & discussion here is a bit off-topic, but the reason why I think it's important to well-define this structure is because docker can eat up a lot of disk space if not configured properly.

Of now what I understand is that we only have one host dockerd running in rootless mode, and we mount it into the github-runner-manager in order to allow it to create isolated container for each run.

If that's the case, then it's OK to me. Just keep in mind that rootless mode may use less performant fs driver that make disk I/O slower and requires more space. In our case we don't use that much space anyway, so maybe we never run into issues.

Reply:
OK thanks for the precision. Then I discover I do not know what is Docker in Docker, do you have a link to explain how to do it properly ?

Reply:
I've just search on google but can't find a good source to explain DinD, it's true that it's not used very often so it's hard to find resources about that.

The only case where I use it is because I have a big docker compose having 10+ containers inside. Imagine running 2 of that such docker compose within the same docker installation, and when doing `docker ps` it will show more active 20+ containers.

The idea of DinD is to spawn a `dockerd` service inside the container (which provides an isolated `docker.sock`), so "real" DinD won't need to mount `host/docker.sock:container/docker.sock`. This way, inside container, `docker ps` only shows containers spawned inside container (10+ in my case).

The tricky part is how to run `dockerd` inside the container, because there is no systemd inside. A trick that I use is simply run `dockerd &` command (so the dockerd run in background).

However, I don't see DinD is needed in our case for now, since we're not running more than 1 container per CI run.